cpp/src/arrow/filesystem/gcsfs.cc

```diff
@@ -536,8 +546,7 @@ class GcsFileSystem::Impl {
                                                            gcs::ReadFromOffset offset) {
     auto stream = client_.ReadObject(path.bucket, path.object, generation, offset);
     ARROW_GCS_RETURN_NOT_OK(stream.status());
-    return std::make_shared<GcsInputStream>(std::move(stream), path, gcs::Generation(),
-                                            offset, client_);
+    return std::make_shared<GcsInputStream>(std::move(stream), path, generation, client_);
```

Comment:
Assume `offset` into this function is 1000.   Without the `offset` parameter passed to `GcsInputStream` its `Tell()` function will return 0 when you are in fact reading byte 1000.  That seems like the wrong semantics to me, but maybe it is the expected behavior?

Reply:
Empirically the underlying tell returns 1000.  I was observing doubling of expected tell value.  The python test that found this wrote N bytes the seemed to N/2 and tried reading.  The reading called the FS tell which returned N

Reply:
I looked at the GCS code and it does appear to keep the ReadAt offset to return for tell but I might have missed something

Reply:
Ack. SGTM.

Reply:
I don't think generic file systems have any notion of generation.  This probably needs to be an enhancement to the GCS unit tests, but I'm not sure test bench supports generation?  @coryan